peer/gossip.go

```diff
@@ -0,0 +1,140 @@
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+	"go.uber.org/zap"
+)
+
+type GossipFunc func(ctx context.Context, subnet id.Hash, contentID []byte) error
+
+// Gossiper returns a new set of callback functions and a gossip function that can be used to spread information throughout a network
+
+func Gossiper(t *transport.Transport, contentResolver dht.ContentResolver, addressTable Table, logger *zap.Logger, next Callbacks) (Callbacks, GossipFunc) {
+
+	gossip := func(ctx context.Context, subnet id.Hash, contentID []byte) error {
+		msg := wire.Msg{Version: wire.MsgVersion1, Type: wire.MsgTypePush, Data: contentID}
+
+		var chainedError error = nil
+		var receivers []id.Signatory
+		if bytes.Compare(subnet[:], GlobalSubnet[:]) == 0 {
+			receivers = addressTable.All()
+		} else {
+			receivers = addressTable.Subnet(subnet)
+		}
+		
+		for _, sig := range receivers {
+			addr, ok := addressTable.PeerAddress(sig)
+			if !ok {
+				logger.Error("gossip", zap.String("table", "peer not found"))
+				continue
+			}
+			if err := t.Send(ctx, sig, addr, msg); err != nil {
+				if chainedError == nil {
+					chainedError = fmt.Errorf("%v, gossiping to peer %v : %v", chainedError, sig, err)
+				} else {
+					chainedError = fmt.Errorf("gossiping to peer %v : %v", sig, err)
+				}
+			}
+		}
+		return chainedError
+	}
+	didReceivePush := func(from id.Signatory, msg wire.Msg) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		addr, ok := addressTable.PeerAddress(from)
+		if !ok {
+			logger.Error("gossip", zap.String("table", "peer not found"))
+			return
+		}
+
+		hash := [32]byte{}
+		copy(hash[:], msg.Data[:])
+
+		if _, ok := contentResolver.Content(hash); !ok {
+			response := wire.Msg{
+				Version: wire.MsgVersion1,
+				Type: wire.MsgTypePull,
+				Data: hash[:],
+			}
+
+			contentResolver.Insert(hash, nil)
+			if err := t.Send(ctx, from, addr, response); err != nil {
+				contentResolver.Delete(hash)
+				logger.Error("gossip", zap.NamedError("pull", err))
+				return
+			}
+
+		}
+	}
+	didReceivePull := func(from id.Signatory, msg wire.Msg) {
+		hash := [32]byte{}
+		copy(hash[:], msg.Data[:])
+		if data, ok := contentResolver.Content(hash); ok {
+			response := wire.Msg{
+				Version: wire.MsgVersion1,
+				Type: wire.MsgTypeSync,
+				Data: data,
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			addr, ok := addressTable.PeerAddress(from)
+			if !ok {
+				logger.Error("gossip", zap.String("table", "peer not found"))
+				return
+			}
+			if err := t.Send(ctx, from, addr, response); err != nil {
+				logger.Error("gossip", zap.NamedError("pull", err))
+			}
+			return
+		}
+
+		logger.Error("gossip", zap.String("pull request", "data not present"))
+	}
+	didReceiveSync := func(from id.Signatory, msg wire.Msg) {
+		hash := id.NewHash(msg.Data)
+		content, ok := contentResolver.Content(dht.ContentID(hash))
```

Comment:
This is incorrect. You cannot assume that the content identifier will be equal to the hash of the content. This is never actually the case. There must be an agreed standard that, in a sync message, there are two bytes slices encoded into the data (using the surge format): one for the content identifier, and the other for the data.
```
| contentID length (n) | contentID  | content length (m) | content  |
| 4 bytes              | n bytes    | 4 bytes            | m bytes  |
```

Reply:
We can achieve this by updating the `wire.Msg` to include a `SyncData` field that is _ignored_ during marshaling and unmarshaling, and then explicitly marshaled/unmarshaled by the Channel when necessary (as per our whiteboard discussion).